### DISCUSSION TITLE
CPU Allocation Irregularities

### DISCUSSION THREAD

**Discussion by valentinhttr** [jobrunr/jobrunr#771] [Q&A]
Hi,

I get the following warning on the Jobrunr dashboard:
```
JobRunr detected some CPU Allocation Irregularities (e.g. due to a very long garbage collect cycle
or due to CPU starvation on shared hosting). This may result in unstable behaviour of your JobRunr cluster.
```

Although this error may seem obvious, I can't seem to correct it. I tried to learn more by reading [the code](https://github.com/jobrunr/jobrunr/blob/master/core/src/main/java/org/jobrunr/server/ServerZooKeeper.java#L142-L153) directly, but it didn't help.

My application is simply hosted via a Spring boot node through Virtuozzo Application Platform (formerly Jelastic). Here is the configuration :
![CleanShot 2023-05-23 at 11 35 29](https://github.com/jobrunr/jobrunr/assets/20521112/364deeb7-78e9-4bab-94cb-40b77197e2bf)

Here are statistics for the last week with 1 hour interval:
![CleanShot 2023-05-23 at 11 40 00](https://github.com/jobrunr/jobrunr/assets/20521112/388a6834-dc4d-41b1-8d0d-faa74240cfe9)
![CleanShot 2023-05-23 at 11 40 31](https://github.com/jobrunr/jobrunr/assets/20521112/71f3d440-718b-4419-bb76-be4514111060)

**Notes**:

- My application is launched without specific java arguments
- Sometimes jobrunr stops by itself...

Do you think this could be due to Vertical Scaling?

Any help would be appreciated 🙏🏻

**Comment by rdehuyss** [comment]
Feel free to join this '[Meet the geek' session](https://www.jobrunr.io/en/blog/2023-05-24-meet-the-geek/) where I'll try to answer your question.

**Reply by valentinhttr** [reply]
Thanks for your proposal, I think it's great to discuss with the community!
But unfortunately, I won't be able to be there. Would you be willing to write me your opinion on the matter here please?

Although I think it's a vertical scaling problem. The readme says:
>You can start small and process jobs within your web app or **scale horizontally** and add as many background job servers as you want to handle a peak of jobs
So I guess vertical scaling is not something I should do.

I tried to put fixed resources and I don't get the warning anymore.

**Reply by rdehuyss** [reply]
Well, this topic alone is so expansive that I can probably easily fill a couple of hours 😄 only with this.

What it basically means is that the JVM where JobRunr is running did not get any CPU cycles at all for a certain period of time. If you're on shared hosting, this may be because other applications may take all CPU sometimes.

Again, this is a topic which is so expansive that it's too much to just write it down here. I'm happy you got it resolved.

**Reply by valentinhttr** [reply]
Ok thanks for your answer. Yes that's what I had read. 
I tried to switch to horizontal scaling with reserved resources, but I have the warning that appeared during the night.

It's strange... I'm not sure why I have this actually, especially with reserved resources, they shouldn't be used by another process.
